### ISSUE TITLE
Rule request - ban `pytest.fixture(autouse=True)`

### ISSUE THREAD

**Issue by Kilo59** [astral-sh/ruff#12491]
I'd like to see a rule that allows for the banning of "autouse" pytest fixtures.

https://docs.pytest.org/en/stable/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request

In a large codebase, the overuse of auto-use fixtures can make reasoning about tests very difficult and cause a lot of unexpected behavior.

This is particularly true for `autouse` fixtures in `conftest.py` files as they just become automatically applied to any tests in scope of the `conftest.py` file.

It would make sense as a `PT` rule.
https://docs.astral.sh/ruff/rules/#flake8-pytest-style-pt

**Comment by jakob1379** [comment]
can we pretty-please have this 😻

**Comment by Kilo59** [comment]
I'm happy to attempt this implementation.
I think we might also want a setting to ban it outright or only ban them in conftest.py files.

**Comment by ntBre** [comment]
This sounds reasonable to me and also straightforward to implement. I think we may want to make this a RUF rule if it doesn't exist upstream in flake8-pytest-style, though. I took a quick look and didn't see an upstream request for it.

> I think we might also want a setting to ban it outright or only ban them in conftest.py files.

This also sounds reasonable to me, but it might be worth trying it without a setting first, unless most people following this issue only want this for conftest files.

**Comment by tjkuson** [comment]
What would the rule recommend doing instead of `autouse=True`? Would it be to add `pytest.mark.usefixtures` to each item?

**Comment by jakob1379** [comment]
> What would the rule recommend doing instead of `autouse=True`? Would it be to add `pytest.mark.usefixtures` to each item?

Depending on if fixes should exist you would need to pass the AST and check which tests include anything from the fixture to come up with good recommendations as `autouse=True` applies to all tests and naively suggesting marks to all methods might not be what you wnat

**Comment by tjkuson** [comment]
Hm, I am not sure if universally suggesting `pytest.mark.usefixtures` as a replacement for `autouse=True` is a good idea (or at least it has downsides). For example, users might forget to apply the marker if it's not auto-used and it could discourage users implementing otherwise helpful fixtures (such as fixtures that guard against test pollution by resetting state between tests). If this rule is added, it'd be neat if it could discriminate between valid and superfluous `autouse=True` usage (not sure how that could be done at the moment).

I'm also curious if people think that fixture scope should play a factor here. For example, I think an auto-used fixture with `scope="session"` has a relatively simple to reason about compared to other pytest concepts (it is setup at the start of a test run and torn down at the end).

**Comment by Kilo59** [comment]
> Hm, I am not sure if universally suggesting `pytest.mark.usefixtures` as a replacement for `autouse=True` is a good idea (or at least it has downsides). For example, users might forget to apply the marker if it's not auto-used and it could discourage users implementing otherwise helpful fixtures (such as fixtures that guard against test pollution by resetting state between tests). If this rule is added, it'd be neat if it could discriminate between valid and superfluous `autouse=True` usage (not sure how that could be done at the moment).
> 
> I'm also curious if people think that fixture scope should play a factor here. For example, I think an auto-used fixture with `scope="session"` has a relatively simple to reason about compared to other pytest concepts (it is setup at the start of a test run and torn down at the end).

@tjkuson 
Yes I agree on both counts.

Always suggesting `pytest.mark.usefixtures` may not be right either.
I think where the autouse fixture is defined and what scope is being used are the best signals for determining how "bad"/problematic it is.

I'm not sure exactly what the line is but; function-scoped, autouse fixtures defined in the root `conftest.py` would be the most egregious and possibly a good default for this rule. With configuration options to tweak that.

Just finished my initial attempt at implementing this. I intentionally have not included any configuration options as suggested by @ntBre, but I'm happy to add them later.

- https://github.com/astral-sh/ruff/pull/25477

**Comment by tjkuson** [comment]
> I think where the autouse fixture is defined and what scope is being used are the best signals for determining how "bad"/problematic it is.

Function-scoped `autouse=True` fixtures in root conftest files seem to be a fairly common pattern to set state between tests ([ref](https://grep.app/search?f.path.pattern=tests%2Fconftest.py&q=autouse%3DTrue)). I don't quite follow what's inherently wrong with autouse or what a better alternative would be (absent a stronger a signal of superfluous usage).

**Comment by ntBre** [comment]
I'm definitely thinking of this as a [pedantic](https://doc.rust-lang.org/stable/clippy/lints.html#pedantic) lint, to use Clippy's terminology. It's restricting a valid pattern in a way that I expect only some projects will want to enforce. I think it's also fine to use this kind of rule with a liberal sprinkling of `noqa` comments, if you do want to enable it in general.

After thinking about it a bit more, I also think it shouldn't be necessary to add a separate configuration option to restrict this to conftest.py files because a glob like this should have the same effect:

```toml
[lint.per-file-ignores]
"!**/conftest.py" = ["RUF076"]
```

**Comment by RonnyPfannschmidt** [comment]
Im a pytest maintainer and I consider this rule highly hostile to sensible usage

**Comment by flying-sheep** [comment]
There are extremely valid use cases for it, I use these all over the place for setup and teardown.

This shouldn’t ever be part of the stable `RUF` group, how can I prevent this?

> […] you may want to ignore it outside of `conftest.py` files, as autouse fixtures are most problematic when defined globally.

Yeah no, I use them in conftest files exactly like elsewhere: for setup and teardown of a large number of tests (conftest files can be used in subdirectories, e.g. you can have `tests/plotting/{conftest.py,…}`)

**Comment by ntBre** [comment]
We're discussing this on https://github.com/astral-sh/ruff/issues/26178, but I think it's likely we will temporarily revert/remove this rule. This isn't part of `RUF` on stable, only on preview, but I didn't consider how common it is to select `RUF` in preview. It will be a better fit to re-add this rule once we have a pedantic lint category after #1774 is resolved.

**Comment by flying-sheep** [comment]
Ah thanks, I should have checked that issue. Indeed that sounds like a great plan!

**Comment by ntBre** [comment]
Reopening as we have temporarily removed the rule again. It was too opinionated/pedantic for the `RUF` category, which many preview users already enable. We can reinstate the rule after #1774, which we are actively working on again.

**Comment by ntBre** [comment]
This rule was added back in 0.16.5, now as part of the preview `pedantic` category rather than `RUF`.

**Comment by RonnyPfannschmidt** [comment]
This rule ought to be removed 

Its simply wrong and against expected pytest usage

**Comment by flying-sheep** [comment]
Yup, see https://github.com/astral-sh/ruff/issues/27959#issuecomment-5485836053

**Comment by ntBre** [comment]
I really don't see why the rule needs to be removed. Some people seem to find it helpful, and we don't recommend enabling it by default.

I do, however, agree with the suggestion to move it to the `restriction` category from `pedantic`.

**Comment by RonnyPfannschmidt** [comment]
Im a pytest core maintainers and the rule encourages patterns that are strictly more error prone and worse
